File: articles/algorithms-in-scientific-computing/fortran/scientific_computing.f90
program scientific_computing
  implicit none
  integer :: i, n
  real(8) :: a, b, h, total, x
  n = 200
  a = 0.0d0
  b = acos(-1.0d0)
  h = (b - a) / dble(n)
  total = 0.5d0 * (sin(a) + sin(b))
  do i = 1, n - 1
     x = a + dble(i) * h
     total = total + sin(x)
  end do
  print *, 'trapezoid_integral=', h * total
end program scientific_computing

File: articles/algorithms-in-scientific-computing/cpp/scientific_computing.cpp
#include <cmath>
#include <iostream>

double f(double x) { return std::sin(x); }
double central_difference(double x, double h) { return (f(x + h) - f(x - h)) / (2.0 * h); }
double trapezoid(int n) {
    const double a = 0.0, b = std::acos(-1.0), h = (b - a) / n;
    double total = 0.5 * (f(a) + f(b));
    for (int i = 1; i < n; ++i) total += f(a + i * h);
    return h * total;
}
int main() {
    std::cout << "central_difference=" << central_difference(1.0, 1e-4) << "\n";
    std::cout << "trapezoid_integral=" << trapezoid(200) << "\n";
}
